base/atomics.jl

```diff
@@ -377,26 +380,26 @@ for typ in atomictypes
                      %irv = extractvalue { $ilt, i1 } %irs, 0
                      %rv = bitcast $ilt %irv to $lt
                      ret $lt %rv
-                     """, $typ, Tuple{Ptr{$typ},$typ,$typ},
+                     """, $typ, Tuple{Ptr{$typ}, $typ, $typ},
                      unsafe_convert(Ptr{$typ}, x), cmp, new)
     end
 
     arithmetic_ops = [:add, :sub]
     for rmwop in [arithmetic_ops..., :xchg, :and, :nand, :or, :xor, :max, :min]
         rmw = string(rmwop)
         fn = Symbol("atomic_", rmw, "!")
-        if (rmw == "max" || rmw == "min") && typ <: Unsigned
+        if (rmw == "max" || rmw == "min") && typ <: Union{Unsigned, Ptr}
             # LLVM distinguishes signedness in the operation, not the integer type.
             rmw = "u" * rmw
         end
         if rmwop in arithmetic_ops && !(typ <: ArithmeticTypes) continue end
-        if typ <: Integer
-            @eval $fn(x::Atomic{$typ}, v::$typ) =
+        if typ <: Union{Integer, Ptr}
+            @eval $fn(x::Atomic{V}, v::V) where {V <: $typ} =
                 llvmcall($"""
                          %ptr = inttoptr i$WORD_SIZE %0 to $lt*
                          %rv = atomicrmw $rmw $lt* %ptr, $lt %1 acq_rel
                          ret $lt %rv
-                         """, $typ, Tuple{Ptr{$typ}, $typ}, unsafe_convert(Ptr{$typ}, x), v)
+                         """, V, Tuple{Ptr{V}, V}, unsafe_convert(Ptr{V}, x), v)
```

Comment:
Here, you'll have to address the fact that `Ptr` doesn't support the same types of `add!` and `sub!` that Integers do:
```julia
julia> Ptr{Int}(0) + Ptr{Int}(1)
ERROR: MethodError: no method matching +(::Ptr{Int64}, ::Ptr{Int64})

julia> Ptr{Int}(0) - Ptr{Int}(1)
0xffffffffffffffff

julia> typeof(Ptr{Int}(0) - Ptr{Int}(1))
UInt64
```

It needs to be handled separately, so that you can add only `atomic_{add!,sub!}(Atomic{<:Ptr}, ::Int)`:
```
[2] +(x::Ptr, y::Integer) in Base at pointer.jl:159
[4] -(x::Ptr, y::Integer) in Base at pointer.jl:160
```

Likely this will have to just be a separate `if` case, rather than combining it with `Integer` as you've done here.
I think you'll want to define two methods: one generic one that converts all `<:Integer` types to `UInt` (as is done here): https://github.com/JuliaLang/julia/blob/74f2de13727373b44da76a65ceb5297889d3e482/base/pointer.jl#L159

And then define the `llvmcall` method for `atomic_add!(Atomic{<:Ptr}, ::UInt)`.

----------------------------------------------------

It is of course too bad that the actual `llvm` string will be the same in both `if`-cases. If you do want to reuse it, I guess you could consider moving it out into a helper variable before this for-loop block, perhaps?

Reply:
The available methods on `Ptr` are a little strange to me. Do you think it isn't worth having `atomic_{add!,sub!}` defined with `Ptr` as the second argument at all?

As for adding/subtracting `<:Integer`, the intrinsics `add_ptr` and `sub_ptr` call LLVM's `getelementptr` which performs pointer arithmetic, although it seems that it's always indexing [elements of type `i8`](https://github.com/JuliaLang/julia/blob/master/src/intrinsics.cpp#L1026). It's not clear to me why this is the case; without looking into it, I might have guessed the arithmetic would increment by `sizeof(T)` for a `Ptr{T}`. 

I wonder if the atomic version should also use `getelementptr` (and then presumably use `atomic_cas!`, similar to some of the current atomic float ops) or just call the `atomicrmw add/sub` instructions. Due to how `add_ptr` and `sub_ptr` are defined at the moment, I think these two methods would be the same only because pointer arithmetic always increments 1 byte at a time.

Reply:
> The available methods on `Ptr` are a little strange to me. Do you think it isn't worth having `atomic_{add!,sub!}` defined with `Ptr` as the second argument at all?

:) Yeah, I can see how it's unintuitive, but no, i think the methods provided by Base are correct, and we should follow them, here. The idea is basically that it is non-meaningful to add two pointers together.

As an analogy, I would say to think of Pointers are basically _points_ on a 1-dimensional line, and Integers as _vectors_ (or _offsets_) on that line. If you subtract two points, you get the vector between them. If you add a vector to a point, you get a new point. And you can add vectors to get a new vector. But it doesn't make sense to add two _points_ together. It's not really even clear what that means, because a point doesn't have a quantity; it's a dimensionless value.

More concretely, `Ptr`s are usually assigned by the kernel's malloc implementation, to some random position in memory. It is almost never going to be the case that adding together two such positions yields a valid position somewhere else in memory. Finally, as one more piece of evidence, for the reasons above, you can't add together pointers in C++ either:
```c++
input_line_6:2:4: error: invalid operands to binary expression ('int *' and 'int *')
 a + b
 ~ ^ ~
```
And similarly, in C++ subtracting two pointers returns a `long` (int).
Does that help clarify things? :) So yes, I think we should not define `atomic_add!(::Ptr, ::Ptr)` or `atomic_sub!(::Ptr, ::Ptr)`.

Reply:
> although it seems that it's always indexing [elements of type `i8`](https://github.com/JuliaLang/julia/blob/master/src/intrinsics.cpp#L1026). It's not clear to me why this is the case; without looking into it, I might have guessed the arithmetic would increment by `sizeof(T)` for a `Ptr{T}`.

Yeah, this surprised me as well when I came to Julia from C++. It's a different decision than is made in C++.
I'm not entirely sure why, but it's the way pointer arithmetic works in julia. If one wants to increment by whole objects, you use `p += sizeof(eltype(p))`. I _think_ the reason is that unlike in C, pointer arithmetic is rarely used in julia to iterate through an array, instead it seems to be more commonly used to offset into fields (bytes) within an element, so the decision was made to make that easier (rather than forcing you to `reinterpret` the pointer to be a `Ptr{Char}` just to index into it, as you would have to do in C/C++). That's just a guess on my part though! :)
Anyway, yeah, we'll want to support the same of course, where we're always indexing elements of `i8`.



> I wonder if the atomic version should also use `getelementptr` (and then presumably use `atomic_cas!`, similar to some of the current atomic float ops) or just call the `atomicrmw add/sub` instructions. Due to how `add_ptr` and `sub_ptr` are defined at the moment, I think these two methods would be the same only because pointer arithmetic currently always increments 1 byte at a time.

Ah, at first I was going to write that I agree with you that it doesn't matter, but no, it'll have to be implemented using `atomicrmw add/sub`. The reason is that the addition must be _atomic_, whereas `getelementptr` + `atomic_cas!` would require _first_ reading the value, then incrementing it, and _then_ assigning the value via `cas!`. But of course it may have changed in between those, which the `cas!` would then overwrite! So it must be `atomicrmw`, I think. :)

In which case, I think probably just sharing the `llvm` string with integer `add/sub` makes sense to me! :)

Reply:
That all makes sense. I don't think contesting the choices for Julia's pointer arithmetic is really the point of this PR anyway, it just took me some time to figure out what was going on.

I'll go ahead and implement `atomic_{add!,sub!}(::Ptr, ::Integer)` using `atomicrmw {add,sub}`. It's equally possible to do so using `atomic_cas!`, but that's likely less efficient (see how this is done for floats [here](https://github.com/JuliaLang/julia/blob/master/base/atomics.jl#L421)) and not necessary in this case.

Reply:
> I'll go ahead and implement `atomic_{add!,sub!}(::Ptr, ::Integer)` using `atomicrmw {add,sub}`. It's equally possible to do so using `atomic_cas!`, but that's likely less efficient (see how this is done for floats [here](https://github.com/JuliaLang/julia/blob/master/base/atomics.jl#L421)) and not necessary in this case.

Oh, cool, makes sense. I hadn't seen that!
But yep, the `atomicrmw` plan sounds great. Lemme know when i should take a look again! :) thanks